File: gherkin/java/src/main/java/gherkin/messages/ParserCucumberMessages.java
package gherkin.messages;

import gherkin.GherkinDocumentBuilder;
import gherkin.Parser;
import gherkin.ParserException;
import gherkin.TokenMatcher;
import gherkin.pickles.PickleCompiler;
import io.cucumber.messages.Messages.Attachment;
import io.cucumber.messages.Messages.GherkinDocument;
import io.cucumber.messages.Messages.Location;
import io.cucumber.messages.Messages.Pickle;
import io.cucumber.messages.Messages.Source;
import io.cucumber.messages.Messages.SourceReference;
import io.cucumber.messages.Messages.Envelope;

import java.util.ArrayList;
import java.util.List;

public class ParserCucumberMessages implements CucumberMessages {
    private final Parser<GherkinDocument.Builder> parser = new Parser<>(new GherkinDocumentBuilder());
    private final TokenMatcher matcher = new TokenMatcher();
    private final PickleCompiler pickleCompiler = new PickleCompiler();

    private final List<String> paths;
    private final boolean includeSource;
    private final boolean includeGherkinDocument;
    private final boolean includePickles;

    public ParserCucumberMessages(List<String> paths, boolean includeSource, boolean includeGherkinDocument, boolean includePickles) {
        this.paths = paths;
        this.includeSource = includeSource;
        this.includeGherkinDocument = includeGherkinDocument;
        this.includePickles = includePickles;
    }

    @Override
    public List<Envelope> messages() {
        List<Envelope> messages = new ArrayList<>();

        FileSources fileSources = new FileSources(paths);
        for (Source source : fileSources) {
            if (includeSource) {
                messages.add(Envelope.newBuilder().setSource(source).build());
            }
            try {
                GherkinDocument gherkinDocument = null;

                if (includeGherkinDocument) {
                    gherkinDocument = buildGherkinDocument(source);
                    messages.add(Envelope.newBuilder().setGherkinDocument(gherkinDocument).build());
                }
                if (includePickles) {
                    if (gherkinDocument == null) {
                        gherkinDocument = buildGherkinDocument(source);
                    }
                    List<Pickle> pickles = pickleCompiler.compile(gherkinDocument, source.getUri());
                    for (Pickle pickle : pickles) {
                        messages.add(Envelope.newBuilder().setPickle(pickle).build());
                    }
                }
            } catch (ParserException.CompositeParserException e) {
                for (ParserException error : e.errors) {
                    addErrorAttachment(messages, error, source.getUri());
                }
            } catch (ParserException e) {
                addErrorAttachment(messages, e, source.getUri());
            }
        }
        return messages;
    }

    private GherkinDocument buildGherkinDocument(Source source) {
        return parser.parse(source.getData(), matcher).setUri(source.getUri()).build();
    }

    private void addErrorAttachment(List<Envelope> messages, ParserException e, String uri) {
        Attachment attachment = Attachment.newBuilder()
                // TODO: Set media here?
                .setSource(SourceReference.newBuilder()
                        .setUri(uri)
                        .setLocation(
                                Location.newBuilder()
                                        .setLine(e.location.getLine())
                                        .setColumn(e.location.getColumn())
                                        .build()
                        )
                        .build())
                .setData(e.getMessage())
                .build();
        messages.add(Envelope.newBuilder().setAttachment(attachment).build());
    }
}

File: gherkin/ruby/lib/gherkin/messages/parser_cucumber_messages.rb
require 'cucumber/messages'
require 'gherkin/parser'
require 'gherkin/pickles/compiler'

module Gherkin
  module Messages
    class ParserCucumberMessages
      def initialize(paths, print_source, print_ast, print_pickles)
        @paths, @print_source, @print_ast, @print_pickles = paths, print_source, print_ast, print_pickles
        @parser = Gherkin::Parser.new
        @compiler = Pickles::Compiler.new
      end
      
      def messages
        Enumerator.new do |y|
          sources.each do |source|
            y.yield(Cucumber::Messages::Wrapper.new(source: source)) if @print_source
            
            begin
              gherkin_document = nil
              
              if @print_ast
                gherkin_document = build_gherkin_document(source)
                y.yield(Cucumber::Messages::Wrapper.new(gherkinDocument: gherkin_document))
              end
              if @print_pickles
                gherkin_document ||= build_gherkin_document(source)
                pickles = @compiler.compile(gherkin_document, source.uri)
                pickles.each do |pickle|
                  y.yield(Cucumber::Messages::Wrapper.new(pickle: pickle))
                end
              end
            rescue CompositeParserException => err
              yield_error_attachments(y, err.errors, source.uri)
            rescue ParserException => err
              yield_error_attachments(y, [err], source.uri)
            end
          end
        end
      end
    
    private

      def yield_error_attachments(y, errors, uri)
        errors.each do |err|
          attachment = Cucumber::Messages::Attachment.new(
            source: Cucumber::Messages::SourceReference.new(
              uri: uri,
              location: Cucumber::Messages::Location.new(
                line: err.location[:line],
                column: err.location[:column]
              )
            ),
            data: err.message
          )
          y.yield(Cucumber::Messages::Wrapper.new(attachment: attachment))
        end
      end

      def sources
        Enumerator.new do |y|
          @paths.each do |path|
            source = Cucumber::Messages::Source.new({
              uri: path,
              data: File.open(path, 'r:UTF-8', &:read),
              media: Cucumber::Messages::Media.new({
                encoding: 'UTF-8',
                content_type: 'text/x.cucumber.gherkin+plain'
              })
            })
            y.yield(source)
          end
        end
      end

      def build_gherkin_document(source)
        gd = @parser.parse(source.data)
        gd[:uri] = source.uri
        Cucumber::Messages::GherkinDocument.new(gd)
      end

    end
  end
end